src/Workspaces/Remote/ServiceHub/Host/AssetProvider.cs

```diff
@@ -284,16 +284,21 @@ Task SynchronizeProjectAssetCollectionAsync<TAsset>(AssetPath assetPath, Func<Pr
                 if (missingChecksumsCount > 0)
                 {
                     var missingChecksumsMemory = new ReadOnlyMemory<Checksum>(missingChecksums, 0, missingChecksumsCount);
+                    Contract.ThrowIfTrue(missingChecksumsMemory.Length == 0);
 
-                    await RequestAssetsAsync(
-                        assetPath, missingChecksumsMemory,
+#if NETCOREAPP
+                    Contract.ThrowIfTrue(missingChecksumsMemory.Span.Contains(Checksum.Null));
+#else
+                    Contract.ThrowIfTrue(missingChecksumsMemory.Span.IndexOf(Checksum.Null) >= 0);
+#endif
+
+                    await _assetSource.GetAssetsAsync(
+                        _solutionChecksum, assetPath, missingChecksumsMemory, _serializerService,
                         static (
-                            int index,
+                            Checksum missingChecksum,
                             T missingAsset,
                             (AssetProvider assetProvider, Checksum[] missingChecksums, Action<Checksum, T, TArg>? callback, TArg? arg) tuple) =>
                         {
-                            var missingChecksum = tuple.missingChecksums[index];
-
                             tuple.callback?.Invoke(missingChecksum, missingAsset, tuple.arg!);
```

Comment:
thsi callback was already "checksum+asset" based.  That gets simpler now since that's what the component we call passes us.